lib/diagnostics_channel.js

```diff
@@ -0,0 +1,281 @@
+'use strict';
+
+const {
+  ArrayPrototypeIndexOf,
+  ArrayPrototypePush,
+  ArrayPrototypeShift,
+  ArrayPrototypeSplice,
+  ObjectCreate,
+  ObjectSetPrototypeOf,
+  Map,
+  MapPrototypeDelete,
+  MapPrototypeGet,
+  MapPrototypeSet,
+  Promise,
+  PromiseResolve,
+  SymbolAsyncIterator
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+const { triggerUncaughtException } = internalBinding('errors');
+
+const { WeakReference } = internalBinding('util');
+
+function identity(v) {
+  return v;
+}
+
+class AsyncIterableChannel {
+  constructor(channel) {
+    this.channel = channel;
+    this.events = [];
+    this.waiting = [];
+
+    this.subscriber = (message) => {
+      const resolve = ArrayPrototypeShift(this.waiting);
+      if (resolve) {
+        return resolve({
+          value: message,
+          done: false
+        });
+      }
+
+      ArrayPrototypePush(this.events, message);
+    };
+
+    channel.subscribe(this.subscriber);
+  }
+
+  [SymbolAsyncIterator]() {
+    return this;
+  }
+
+  return() {
+    const data = { done: true };
+    this.done = true;
+
+    this.channel.unsubscribe(this.subscriber);
+
+    for (let i = 0; i < this.waiting.length; i++) {
+      const resolve = this.waiting[i];
+      resolve(data);
+    }
+
+    return PromiseResolve(data);
+  }
+
+  next() {
+    const event = ArrayPrototypeShift(this.events);
+    if (event) {
+      return PromiseResolve({
+        value: event,
+        done: false
+      });
+    }
+
+    if (this.done) {
+      return PromiseResolve({
+        done: true
+      });
+    }
+
+    return new Promise((resolve) => {
+      ArrayPrototypePush(this.waiting, resolve);
+    });
+  }
+}
+
+class SpanMessage {
+  constructor(action, id, data) {
+    this.action = action;
+    this.id = id;
+    this.data = data;
+  }
+
+  // Augment with type: span to identify over the wire through JSON
+  toJSON() {
+    return {
+      ...this,
+      type: 'span'
+    };
+  }
+}
+
+let spanId = 0;
+
+class InactiveSpan {
+  message() {}
+  annotate() {}
+  end() {}
+}
+
+class Span {
+  constructor(channel, data) {
+    this.channel = channel;
+    this.id = ++spanId;
+    this.ended = false;
+
+    this.message('start', data);
+  }
+
+  message(action, data) {
```

Comment:
```suggestion
  _message(action, data) {
```
I think this is a private function

Reply:
Yes, though considering making it public so a user could just send stuff with whatever action name they want. 🤔

Reply:
This would require that a user knows also the special meaning/names of the `action`s.

Reply:
The only "special" name is "end" which is only special in the context of the `aggregate` method. The names can be literally anything. The start/annotate/end included by default are purely informational, they could be whatever you want them to be.

Reply:
The publisher has to follow the span semantics but it doesn't know if the consumers use `aggregate` or not.
If end is omitted `aggregate` will leak.